vm/transaction.go

```diff
@@ -58,3 +50,171 @@ func marshalTxn(txn core.Transaction) (json.RawMessage, error) {
+	MaxFee                *felt.Felt                   `json:"max_fee,omitempty"`
+	Signature             *[]*felt.Felt                `json:"signature,omitempty"`
+	CallData              *[]*felt.Felt                `json:"calldata,omitempty"`
+	EntryPointSelector    *felt.Felt                   `json:"entry_point_selector,omitempty"`
+	Nonce                 *felt.Felt                   `json:"nonce,omitempty"`
+	CompiledClassHash     *felt.Felt                   `json:"compiled_class_hash,omitempty"`
+	ResourceBounds        *map[Resource]ResourceBounds `json:"resource_bounds,omitempty"`
+	Tip                   *felt.Felt                   `json:"tip,omitempty"`
+	NonceDAMode           *DataAvailabilityMode        `json:"nonce_data_availability_mode,omitempty"`
+	FeeDAMode             *DataAvailabilityMode        `json:"fee_data_availability_mode,omitempty"`
+	AccountDeploymentData *[]*felt.Felt                `json:"account_deployment_data,omitempty"`
+	PaymasterData         *[]*felt.Felt                `json:"paymaster_data,omitempty"`
+}
+
+type DataAvailabilityMode uint32
+
+const (
+	DAModeL1 DataAvailabilityMode = iota
+	DAModeL2
+)
+
+func (m DataAvailabilityMode) MarshalJSON() ([]byte, error) {
+	switch m {
+	case DAModeL1:
+		return []byte(`"L1"`), nil
+	case DAModeL2:
+		return []byte(`"L2"`), nil
+	default:
+		return nil, errors.New("unknown resource")
+	}
+}
+
+type Resource uint32
+
+const (
+	ResourceL1Gas Resource = iota + 1
+	ResourceL2Gas
+)
+
+func (r Resource) MarshalText() ([]byte, error) {
+	switch r {
+	case ResourceL1Gas:
+		return []byte("L1_GAS"), nil
+	case ResourceL2Gas:
+		return []byte("L2_GAS"), nil
+	default:
+		return nil, errors.New("unknown resource")
+	}
+}
+
+func (r Resource) MarshalJSON() ([]byte, error) {
+	result, err := r.MarshalText()
+	if err != nil {
+		return nil, err
+	}
+	return []byte(`"` + string(result) + `"`), nil
+}
+
+type ResourceBounds struct {
+	MaxAmount       *felt.Felt `json:"max_amount"`
+	MaxPricePerUnit *felt.Felt `json:"max_price_per_unit"`
+}
+
+func adaptTransaction(txn core.Transaction) *Transaction {
+	var tx *Transaction
+	switch t := txn.(type) {
+	case *core.DeclareTransaction:
+		tx = &Transaction{
+			Hash:              t.Hash(),
+			MaxFee:            t.MaxFee,
+			Version:           t.Version.AsFelt(),
+			Signature:         utils.Ptr(t.Signature()),
+			Nonce:             t.Nonce,
+			ClassHash:         t.ClassHash,
+			SenderAddress:     t.SenderAddress,
+			CompiledClassHash: t.CompiledClassHash,
+		}
+
+		if tx.Version.Uint64() == 3 {
+			tx.ResourceBounds = utils.Ptr(adaptResourceBounds(t.ResourceBounds))
+			tx.Tip = new(felt.Felt).SetUint64(t.Tip)
+			tx.PaymasterData = &t.PaymasterData
+			tx.AccountDeploymentData = &t.AccountDeploymentData
+			tx.NonceDAMode = utils.Ptr(DataAvailabilityMode(t.NonceDAMode))
+			tx.FeeDAMode = utils.Ptr(DataAvailabilityMode(t.FeeDAMode))
+		}
+	case *core.InvokeTransaction:
```

Comment:
Do we need to special-case invoke v0s?

Reply:
https://github.com/NethermindEth/juno/pull/1488/files#diff-a060f37d0efb55c39bb433ff6ccaa6116ac7fa1462a49d28c64719a010e16373L32-L36 was a workaround that we don't need any more.